modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/ElasticsearchIndexSearcher.java

```diff
@@ -459,6 +504,177 @@ private void _prepare(
+				break;
+			}
+
+			int[] startAndEnd = SearchPaginationUtil.calculateStartAndEnd(
+				start, end, hits.getLength());
+
+			start = startAndEnd[0];
+			end = startAndEnd[1];
+		}
+
+		return hits;
+	}
+
+	private Hits _searchWithDeepPagination(
+		int end, Query query, SearchContext searchContext,
+		SearchRequest searchRequest,
+		SearchResponseBuilder searchResponseBuilder, int start) {
+
+		SearchSearchRequest searchSearchRequest =
+			_createSearchSearchRequestWithDeepPagination(
+				searchRequest, searchContext, query);
+
+		SearchSearchResponse searchSearchResponse = null;
+
+		int maxResultWindow = 10000; // GL: should be configurable
+
+		try {
+			if (end > maxResultWindow) {
+				int documentsToSkip = 0;
+
+				// GL: This validation is to be able to return results between
+				// the maxResultWindow limit for example: [9999 to 10001]
+
+				searchSearchRequest.setSize(1);
+
+				if (start < maxResultWindow) {
+					searchSearchRequest.setStart(start - 1);
+				}
+				else {
+					searchSearchRequest.setStart(maxResultWindow - 1);
+
+					documentsToSkip = start % maxResultWindow;
+				}
+
+				// GL: first search could be just the 10000th document
+				// or if the start is lower than 10000, the first search could
+				// be the document before the start position
+
+				searchSearchResponse = _searchEngineAdapter.execute(
+					searchSearchRequest);
+
+				int maxResultWindowPages = start / maxResultWindow;
+
+				// GL: for each maxResultWindow before the start position,
+				// we need to execute a search to get the last document
+				// of the last page
+
+				for (int i = 1; i < maxResultWindowPages; i++) {
+					SearchHit lastSearchHit = _getLastSearchHit(
+						searchSearchResponse);
+
+					if (lastSearchHit == null) {
+						return new HitsImpl();
+					}
+
+					searchSearchRequest.setSearchAfter(
+						lastSearchHit.getSortValues());
+					searchSearchRequest.setSize(maxResultWindow);
+					searchSearchRequest.setStart(0);
+
+					searchSearchResponse = _searchEngineAdapter.execute(
+						searchSearchRequest);
+				}
+
+				SearchHit lastSearchHit = _getLastSearchHit(
+					searchSearchResponse);
+
+				if (lastSearchHit == null) {
+					return new HitsImpl();
+				}
+
+				searchSearchRequest.setSearchAfter(
+					lastSearchHit.getSortValues());
+
+				// GL: Because search after need to have start = 0
+				// we maybe need to skip some documents to get the correct
+				// start position
+
+				if (documentsToSkip > 0) {
+					searchSearchRequest.setSize(documentsToSkip);
+					searchSearchRequest.setStart(0);
+
+					searchSearchResponse = _searchEngineAdapter.execute(
+						searchSearchRequest);
+
+					lastSearchHit = _getLastSearchHit(searchSearchResponse);
+
+					if (lastSearchHit == null) {
+						return new HitsImpl();
+					}
+
+					searchSearchRequest.setSearchAfter(
+						lastSearchHit.getSortValues());
+				}
+
+				searchSearchRequest.setStart(0);
+			}
+
+			searchSearchRequest.setStart(start);
```

Comment:
It should be inside an else statement but SF removed it, that's why it probably is not working in your machine. I will resend it.

Reply:
https://github.com/liferay/liferay-portal/blob/master/modules/util/source-formatter/src/main/resources/documentation/check/java_else_statement_check.markdown#javaelsestatementcheck

Reply:
It is because of:

`
if (lastSearchHit == null) {
					return new HitsImpl();
				}
`